### PR TITLE
fix(datasource/docker): 405 for auth and no HttpError

### PR DESCRIPTION
**Pull request by chezsmithy**
## Changes

Allowed authorization for docker registries to support 405 as a successful result along with 200.
Allowed the request for tags to support both an HttpError or a success result when the upstream ECR repository from the un-authenticated public proxy returns a 405.

## Context

We are using an ECR proxy fronted with a public and open proxy server.
When attempting to pull tags from the proxy, with tags 10000 we received the 405 as expected. Unfortunately this was unexpected for the auth call and failed. Further when auth allowed 405, as success with no headers, the next call to obtain tags without an Auth Header did not result in an HttpError. Thus the backoff to 1000 tags failed.

Note that attempting to write a test for the 405 result without an HttpError is proving extremely difficult. As such the code coverage has been affected.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [X] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [] Code inspection only, or
- [] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [X] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/docker/index.ts [hunk 1]
```diff
@@ -70,15 +67,13 @@ export async function getAuthHeaders(
...
+      apiCheckResponse.statusCode === 405
+    ) {
       logger.debug({ apiCheckUrl }, 'No registry auth required');
       return {};
     }
-    if (apiCheckResponse.statusCode === 404) {
-      logger.debug({ apiCheckUrl }, 'Page Not Found');
-      // throw error up to be caught and potentially retried with library/ prefix
-      throw new Error(PAGE_NOT_FOUND_ERROR);
-    }
```
Comment (viceice) [inline lib/modules/datasource/docker/index.ts:81]: definitely No-Go

File: lib/modules/datasource/docker/index.ts [hunk 2]
```diff
@@ -70,15 +67,13 @@ export async function getAuthHeaders(
         // TODO: add cache test
         await http.getJson(apiCheckUrl, options);
 
-    if (apiCheckResponse.statusCode === 200) {
+    if (
+      apiCheckResponse.statusCode === 200 ||
+      apiCheckResponse.statusCode === 405
```
Comment (viceice) [inline lib/modules/datasource/docker/index.ts:72]: should only be allowed when apicheckurl doesn't end with `/v2/`.

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/datasource/docker/index.ts (hunk 2 above)
Comment (chezsmithy) [reply lib/modules/datasource/docker/index.ts:75]: @viceice in our case apicheckUrl is https://ecr-proxy.company.com/v2/node/tags/list?n=10000 and we are getting a 405 response from the un-authenticated proxy from the gttp.getJson Call.

Comment (chezsmithy) [reply lib/modules/datasource/docker/index.ts:75]: @viceice due to the fact that the proxy is handling the authentication for us to AWS ECR.

Comment (chezsmithy) [reply lib/modules/datasource/docker/index.ts:75]: Not changing given my investigation.

File: lib/modules/datasource/docker/index.ts [hunk 3]
```diff
@@ -834,7 +851,7 @@ export class DockerDatasource extends Datasource {
       return null;
     }
     let page = 1;
-    let foundMaxResultsError = false;
+    let foundMaxResults = false;
```
Comment (viceice) [inline lib/modules/datasource/docker/index.ts:854]: don't change variable without goot reason

File: lib/modules/datasource/docker/index.ts [hunk 4]
```diff
@@ -844,18 +861,25 @@ export class DockerDatasource extends Datasource {
...
           const maxResults = 1000;
           url = `${registryHost}/${dockerRepository}/tags/list?n=${maxResults}`;
           url = ensurePathPrefix(url, '/v2');
-          foundMaxResultsError = true;
+          foundMaxResults = true;
           continue;
         }
         throw err;
       }
+      if (!foundMaxResults && isECRMaxResultsResponse(res)) {
```
Comment (viceice) [inline lib/modules/datasource/docker/index.ts:876]: This is totally wrong, as res is undefined or the result of the prevous request. So revert

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/datasource/docker/index.ts (hunk 3 above)
Comment (chezsmithy) [reply lib/modules/datasource/docker/index.ts:854]: Reverted variable name change.

**Comment by rarkins** [comment]
@chezsmithy do you have the time to move this forward again?

**Comment by rarkins** [comment]
Closing due to inactivity. Welcome to reopen or recreate in future

**Pull request closed without merging**